toolchain/parser/parser_impl.cpp

```diff
@@ -356,16 +361,45 @@ auto ParseTree::Parser::ParseList(TokenKind open, TokenKind close,
 
 auto ParseTree::Parser::ParsePattern(PatternKind kind) -> llvm::Optional<Node> {
   CARBON_RETURN_IF_STACK_LIMITED(llvm::None);
-  if (NextTokenIs(TokenKind::Identifier()) &&
+  auto start = GetSubtreeStartPosition();
+
+  auto continue_parsing = [&] {
```

Comment:
Given that this function can handle various different kinds of pattern, I think this name isn't specific enough. How about something like `finish_parsing_pattern_binding`?